openff/interchange/components/toolkit.py

```diff
@@ -105,35 +108,98 @@ def _check_electrostatics_handlers(force_field: "ForceField") -> bool:
-        )
+        if atom.element is None:
+            assert isinstance(
+                system,
+                openmm.System,
+            ), "`system` argument required if virtual sites are present in the topology"
+            try:
+                # assume ThreeParticleAverageSite for now
+                orientation_atom_indices = [
+                    openmm_openff_particle_map[system.getVirtualSite(atom.index).getParticle(i)] for i in range(3)
+                ]
+            except openmm.OpenMMException as error:
+                if "This particle is not a virtual site" in str(error):
+                    raise ValueError(
+                        "Particle ordering mismatch between OpenMM system and topology. "
+                        f"Look at particle {atom.index} in the topology.",
+                    ) from error
+            virtual_sites.append(
+                ImportedVirtualSiteKey(
+                    orientation_atom_indices=orientation_atom_indices,
+                    name=atom.name,
+                    type="ThreeParticleAverageSite",
+                ),
+            )
+
+            # TODO: This will break if virtual sites are not after their parent/orientation atoms
+            openmm_openff_particle_map[atom.index] = virtual_sites[-1]
+
+        else:
+            graph.add_node(
+                atom.index,
+                atomic_number=getattr(atom.element, "atomic_number", 0),
+                name=atom.name,
+                residue_name=atom.residue.name,
+                # Note that residue number is mapped to residue.id here. The use of id vs. number
+                # varies in other packages and the convention for the OpenFF-OpenMM interconversion
+                # is recorded at
+                # https://docs.openforcefield.org/projects/toolkit/en/0.15.1/users/molecule_conversion.html
+                residue_number=atom.residue.id,
+                insertion_code=atom.residue.insertionCode,
+                chain_id=atom.residue.chain.id,
+            )
+
+            openmm_openff_particle_map[atom.index] = atom.index - len(virtual_sites)
 
     for bond in openmm_topology.bonds():
         graph.add_edge(
             bond.atom1.index,
             bond.atom2.index,
         )
 
-    return _simple_topology_from_graph(graph)
+    topology = _simple_topology_from_graph(graph)
+
+    topology._molecule_virtual_site_map = defaultdict(list)
+
+    # TODO: This iteration strategy scales horribly with system size - need to refactor -
+    #       since looking up topology atom indices is slow. It's probably repetitive to
+    #       look up the molecule index over and over again
+    for particle in virtual_sites:
+        molecule_index = topology.molecule_index(topology.atom(particle.orientation_atom_indices[0]).molecule)
+
+        topology._molecule_virtual_site_map[molecule_index].append(particle)
+
+    topology._particle_map = openmm_openff_particle_map
```

Comment:
Would it be reasonable to squish `topology._particle_map` and `topology._molecule_virtual_site_map` into atom metadata rather than monkey patching? I think the answer is probably "no", and I see this information is moved to the `Interchange`, and I've tested that `Interchange.topology` seems to be able to be swapped out for a true topology (with full molecular graphs etc).

Reply:
Possibly; I don't love the current design but it has the benefit of not interacting with anything the user might do in the public API